### ISSUE TITLE
Allow custom source for CAPI integration

### ISSUE THREAD

**Issue by bmattb** [GoogleForCreators/web-stories-wp#8075]
## Feature Description

Allow custom source for CAPI (NewsCorp DAM) integration.

- Allow filtering providers so we can inject custom for client’s DAM integration.
- Make the provider aware about endpoints which it will use to send requests out instead of generic endpoints.
- [Proof of concept](https://github.com/google/web-stories-wp/compare/main...PatelUtkarsh:extend/provider#diff-bd41bd9aa46fc963f359aa90dffe68c46d64c79c7129100496b72fdffed2d519R95-R138)

## Alternatives Considered

- Integrate into WP media library instead, similar to the Unsplash WP plugin

## Additional Context

- Are all “filters” (photo, video, GIF) omni-search across all providers by design?
  - If “omni” is the approach, can we filter to select only one provider?
- We intend to allow for multiple registrations (e.g. our custom API, possibly to Getty directly, etc.). We intend to override the providerConfiguration.js file and build routing from there. Is this the design of the plugin? 
- What is the easiest way to do this today? (We recognize this isn’t yet possible)
- We also may need to de-register some of the default integrations for licensing reasons, how should we go about this?

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by swissspidy** [comment]
Just sharing my answers to these questions from the [meeting doc](https://docs.google.com/document/d/1xVdrzDtc7MPXiD-GADIK4Kb7X67t2bTHHxIU0-yetAE/edit#) again here for visibility:

> * Are all “filters” (photo, video, GIF) omni-search across all providers by design?
>   
>   * If “omni” is the approach, can we filter to select only one provider?

This is not an omni-search; these are the actual provider selectors.
Images = Unsplash
Video = Coverr
GIFs = Tenor

> * Allow filtering providers so we can inject custom for client’s DAM integration.
> * Make the provider aware about endpoints which it will use to send requests out instead of generic endpoints.
> * We intend to allow for multiple registrations (e.g. our custom API, possibly to Getty directly, etc.). We intend to override the providerConfiguration.js file and build routing from there. Is this the design of the plugin?
> * What is the easiest way to do this today? (We recognize this isn’t yet possible)

The third-party media tab was built so that all these providers are fetched from https://media3p.googleapis.com. It was not really designed to support other API endpoints, hence this is not possible today.

> * We also may need to de-register some of the default integrations for licensing reasons, how should we go about this?

This is not easily possible today, but could probably be achieved using some server-side filters.

This could be a separate ticket on its own.

**Comment by PatelUtkarsh** [comment]
For CAPI DAM integration we are currently looking at integrating it with the media library using either of the following:
1. https://github.com/xwp/unsplash-wp
1. https://github.com/humanmade/asset-manager-framework/

This allows us to maintain our own media sources to be included by users of the Web Stories plugin. 

We are also looking at integrating other source providers like Getty which may follow the same approach as DAM.
Our goal is to create a standard way to allow sources to provide media in the creation of Web Stories. 

We are trying to avoid us (and others) from having to fork the plugin or add a pull request each time a new media source is added. In short, each WP instance should be able to set its media sources for Web Stories.

**Comment by dreamofabear** [comment]
Awesome, that sounds great. Let us know if you find any technical roadblocks when integrating into the WP media library.

**Comment by spacedmonkey** [comment]
I am have been considering this problem and I have worked on a [POC](https://github.com/google/web-stories-wp/tree/try/filter). 

See these screenshots.

<img width="334" alt="Screenshot 2021-07-05 at 14 14 32" src="https://user-images.githubusercontent.com/237508/124479591-55dc3c00-dd9e-11eb-8c45-2c498a4921b2.png">

<img width="1904" alt="Screenshot 2021-07-05 at 14 15 21" src="https://user-images.githubusercontent.com/237508/124478743-6809aa80-dd9d-11eb-937d-803b4caf4733.png">

Basically, what this POC does, it makes the list of filterable media types options in the media library, filterable in PHP. Meaning using the `web_stories_editor_settings` filter, you could add your own type. In my example it is custom. A custom name can even be give, like dam integration or whatever you want. Then the last thing what you need to do, is to extend the `Stories_Media_Controller` / `WP_REST_Attachments_Controller` REST API controller, that when you pass the type of custom ( what whatever the key is) is passed to that controller, you do whatever custom logic is required to load data from third party source. As long as the API response is the same format as the custom controller, it should just work. 

This is extremely easy for the web stories team to implement ( as I have basically already done it) and a small foot print in change ( it is like 7 lines changed, mostly moving code from js to php ). 

Thoughts?

**Comment by PatelUtkarsh** [comment]
Hello @spacedmonkey, 

Your approach looks good to me.

**Comment by dreamofabear** [comment]
@spacedmonkey FYI for DAM integrations, we've been recommending **not** to interface directly with the `web-stories-wp` codebase but rather integrate with the WP media library instead. 

This is because we don't expect to ship any DAM integrations directly in the Web Stories for WordPress plugin (it's a custom use case that won't work for all of our users). This also decouples DAM-related code from our codebase, release cycle, etc.

**Comment by spacedmonkey** [comment]
> @spacedmonkey FYI for DAM integrations, we've been recommending not to interface directly with the web-stories-wp codebase but rather integrate with the WP media library instead.

Yes, I know, I have been advising the best work around. I advised using [asset-manager-framework](https://github.com/humanmade/asset-manager-framework), which can be used to completely replace the WP media library with a third party library. In the future, we should push others down this route. However, AMF, is only useful want to wholesale replace the WP media library with your own. This does not handle the user case where you have two or more sources of media. Take for an example ( on a project I worked on ), a site that used the WP Media library for local assets, a third party dam for images and another for video. AMF would not work here. 

As for my patch, all it really does is allow for a dropdown to be extends to add another option. Unless you write code to integrate with the filter, then it does nothing. But for the people that need this functionality it could unblock them from using this plugin with they editorial workflow. As this is such a small ask, why not unblock users from using our editor?
